### PR TITLE
Add Java7 grammar

### PR DESCRIPTION
**Pull request by ptersilie**
This updates our Java5 grammar to Java7.

### CODE REVIEW

File: java7/java.y [hunk 1]
```diff
@@ -592,8 +592,19 @@ catches :    catch_clause
...
+
+catch_formal_parameter :
+        modifiers_opt catch_type "IDENTIFIER"
+    ;
+
+catch_type :
+        name
+    |   catch_type "OR" name
+    ;
+
```
Comment (ltratt) [inline java7/java.y:676]: Double blank line.

Comment (ltratt) [inline java7/java.y:676]: Double blank line.

File: java7/java.l [hunk 2]
```diff
@@ -98,13 +98,13 @@ instanceof INSTANCEOF
 \|= OREQ
 
 [0-9]+\.[0-9]*([eE][+-]?[0-9]+)?[fFdD]?|\.[0-9]+([eE][+-]?[0-9]+)?[fFdD]?|[0-9]+[eE][+-]?[0-9]+[fFdD]?|[0-9]+[fFdD] FLOATING_POINT_LITERAL
-(0_[0-7]+|0b[01]+|0x[0-9A-Fa-f]+|[0-9]+)[lL]? INTEGER_LITERAL
+(0_[0-7]+|[0-9][0-9_]+[0-9]|0b[01]+|0[xX][0-9A-Fa-f]+|[0-9]+)[lL]? INTEGER_LITERAL
```
Comment (ltratt) [inline java7/java.l]: Shouldn't this be `0[bB][01]+`?

File: java7/java.l [hunk 3]
```diff
@@ -98,13 +98,13 @@ instanceof INSTANCEOF
 \|= OREQ
 
 [0-9]+\.[0-9]*([eE][+-]?[0-9]+)?[fFdD]?|\.[0-9]+([eE][+-]?[0-9]+)?[fFdD]?|[0-9]+[eE][+-]?[0-9]+[fFdD]?|[0-9]+[fFdD] FLOATING_POINT_LITERAL
-(0_[0-7]+|0b[01]+|0x[0-9A-Fa-f]+|[0-9]+)[lL]? INTEGER_LITERAL
+(0_[0-7]+|[0-9][0-9_]+[0-9]|0b[01]+|0[xX][0-9A-Fa-f]+|[0-9]+)[lL]? INTEGER_LITERAL
 (true|false) BOOLEAN_LITERAL
 null NULL_LITERAL
-[a-zA-Z_][a-zA-Z0-9_]* IDENTIFIER
+[a-zA-Z_$][a-zA-Z0-9_$]* IDENTIFIER
```
Comment (ltratt) [inline java7/java.l]: Can an identifier start with `$`?

File: java7/java.y [hunk 4]
```diff
@@ -583,8 +583,33 @@ synchronized_statement :
...
+    |   resource_list "SEMICOLON" resource
+    ;
+
+resource:
+        annotation "FINAL" type variable_declarator_id "EQ" expression
+    |   annotation         type variable_declarator_id "EQ" expression
+    |              "FINAL" type variable_declarator_id "EQ" expression
+    |                      type variable_declarator_id "EQ" expression
+    ;
+
```
Comment (ltratt) [inline java7/java.y:657]: Double blank line.

**Comment by ltratt** [comment]
A few minor comments, but overall this looks good!

File: java7/java.l (hunk 2 above)
Comment (ptersilie) [reply java7/java.l]: Good spot. I only fixed the ones that flagged up when parsing the standard library.

File: java7/java.l (hunk 3 above)
Comment (ptersilie) [reply java7/java.l]: According to the spec, yes: https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8

Comment (ltratt) [reply java7/java.l]: I didn't know that!

Comment (ptersilie) [reply java7/java.l]: That wasn't the only thing that surprised me about the Java specification.

**Comment by ptersilie** [comment]
All fixed.

**Comment by ltratt** [comment]
Looks good to me! If you squash, I can merge.

**Comment by ptersilie** [comment]
Squashed!

**Pull request merged**